Smt/Commands.lean

```diff
@@ -31,52 +38,42 @@ def defNatSub : Command :=
     (`"ite" • (`"<" • `"x" • `"y") • ``"0" • (`"-" • `"x" • `"y"))
     false
 
-def sortEndsWithNat : Term → Bool
-  | arrowT _ t    => sortEndsWithNat t
-  | symbolT "Nat" => true
-  | _             => false
+open ToSexp in
+protected def toSexp : Command → Sexp
+  | .setLogic l                   => sexp!{(set-logic {l})}
+  | .setOption k v                => sexp!{(set-option {s!":{k}"} {v})}
+  | .assert val                   => sexp!{(assert {toSexp val})}
+  | .declare nm st@(arrowT ..)    =>
+    let sts := arrowToList st
+    sexp!{(declare-fun {quoteSymbol nm} (...{sts.init.map toSexp}) {toSexp sts.getLast!})}
+  | .declare nm st                => sexp!{(declare-const {quoteSymbol nm} {toSexp st})}
+  | .declareSort nm arity         =>
+    sexp!{(declare-sort {quoteSymbol nm} {String.mk (Nat.toDigits 10 arity)})}
+  | .defineSort nm ps tm          =>
+    sexp!{(define-sort {quoteSymbol nm} (...{ps.map toSexp}) {toSexp tm})}
+  | .defineFun nm [] cod tm _     =>
+    sexp!{(define-const {quoteSymbol nm} {toSexp cod} {toSexp tm})}
+  | .defineFun nm ps cod tm false =>
+    sexp!{(define-fun {quoteSymbol nm} {paramsToSexp ps} {toSexp cod} {toSexp tm})}
+  | .defineFun nm ps cod tm true  =>
+    sexp!{(define-fun-rec {quoteSymbol nm} {paramsToSexp ps} {toSexp cod} {toSexp tm})}
+  | .checkSat                     => sexp!{(check-sat)}
+  | .getModel                     => sexp!{(get-model)}
+  | .exit                         => sexp!{(exit)}
+where
+  arrowToList : Term → List Term
+    | arrowT d c => d :: arrowToList c
+    | s          => [s]
+  paramToSexp (p : String × Term) : Sexp := sexp!{({quoteSymbol p.fst} {toSexp p.snd})}
+  paramsToSexp (ps : List (String × Term)) : Sexp := sexp!{(...{ps.map paramToSexp})}
 
-def natAssertBody (t : Term) : Term :=
-  mkApp2 (symbolT ">=") t (literalT "0")
+instance : ToSexp Command := ⟨Command.toSexp⟩
 
-variable [Monad m] [MonadNameGenerator m]
+instance : ToString Command := ⟨toString ∘ ToSexp.toSexp⟩
 
-/-- TODO: refactor to support functions as input (e.g., (Nat → Nat) → Nat). -/
-def natConstAssert (n : String) (args : List Name) : Term → m Term
-  | arrowT i@(symbolT "Nat") t => do
-    let id ← Lean.mkFreshId
-    return (forallT id.toString i
-                   (imp id.toString (← natConstAssert n (id::args) t)))
-  | arrowT a t => do
-    let id ← Lean.mkFreshId
-    return (forallT id.toString a (← natConstAssert n (id::args) t))
-  | _ => pure $ natAssertBody (applyList n args)
-  where
-    imp n t := appT (appT (symbolT "=>") (natAssertBody (symbolT n))) t
-    applyList n : List Name → Term
-      | [] => symbolT n
-      | t :: ts => appT (applyList n ts) (symbolT t.toString)
+instance : ToString (List Command) := ⟨.intercalate "\n" ∘ (.map toString) ∘ .reverse⟩
```

Comment:
This instance feels a bit too magic, it also conflicts with the umbrella `[ToString a] : ToString (List a)` instance.

Reply:
I knew I couldn't get away with it 😅

Reply:
Can't come up with a good name. Open for suggestions!

Reply:
`cmdsAsQuery` sounds good 🤷